src/core/ext/filters/client_channel/lb_policy/grpclb/client_load_reporting_filter.cc

```diff
@@ -33,118 +33,49 @@
 #include "src/core/lib/transport/transport.h"
 
-static grpc_error_handle clr_init_channel_elem(
-    grpc_channel_element* /*elem*/, grpc_channel_element_args* /*args*/) {
-  return absl::OkStatus();
-}
-
-static void clr_destroy_channel_elem(grpc_channel_element* /*elem*/) {}
-
-namespace {
-
-struct call_data {
-  // Stats object to update.
-  grpc_core::RefCountedPtr<grpc_core::GrpcLbClientStats> client_stats;
-  // State for intercepting send_initial_metadata.
-  grpc_closure on_complete_for_send;
-  grpc_closure* original_on_complete_for_send;
-  bool send_initial_metadata_succeeded = false;
-  // State for intercepting recv_initial_metadata.
-  grpc_closure recv_initial_metadata_ready;
-  grpc_closure* original_recv_initial_metadata_ready;
-  bool recv_initial_metadata_succeeded = false;
-};
+namespace grpc_core {
+const grpc_channel_filter ClientLoadReportingFilter::kFilter =
+    MakePromiseBasedFilter<ClientLoadReportingFilter, FilterEndpoint::kClient,
+                           kFilterExaminesServerInitialMetadata>(
+        "client_load_reporting");
 
-}  // namespace
-
-static void on_complete_for_send(void* arg, grpc_error_handle error) {
-  call_data* calld = static_cast<call_data*>(arg);
-  if (error.ok()) {
-    calld->send_initial_metadata_succeeded = true;
-  }
-  grpc_core::Closure::Run(DEBUG_LOCATION, calld->original_on_complete_for_send,
-                          error);
+absl::StatusOr<ClientLoadReportingFilter> ClientLoadReportingFilter::Create(
+    const ChannelArgs&, ChannelFilter::Args) {
+  return ClientLoadReportingFilter();
 }
 
-static void recv_initial_metadata_ready(void* arg, grpc_error_handle error) {
-  call_data* calld = static_cast<call_data*>(arg);
-  if (error.ok()) {
-    calld->recv_initial_metadata_succeeded = true;
-  }
-  grpc_core::Closure::Run(DEBUG_LOCATION,
-                          calld->original_recv_initial_metadata_ready, error);
-}
-
-static grpc_error_handle clr_init_call_elem(
-    grpc_call_element* elem, const grpc_call_element_args* args) {
-  GPR_ASSERT(args->context != nullptr);
-  new (elem->call_data) call_data();
-  return absl::OkStatus();
-}
-
-static void clr_destroy_call_elem(grpc_call_element* elem,
-                                  const grpc_call_final_info* /*final_info*/,
-                                  grpc_closure* /*ignored*/) {
-  call_data* calld = static_cast<call_data*>(elem->call_data);
-  if (calld->client_stats != nullptr) {
-    // Record call finished, optionally setting client_failed_to_send and
-    // received.
-    calld->client_stats->AddCallFinished(
-        !calld->send_initial_metadata_succeeded /* client_failed_to_send */,
-        calld->recv_initial_metadata_succeeded /* known_received */);
+ArenaPromise<ServerMetadataHandle> ClientLoadReportingFilter::MakeCallPromise(
+    CallArgs call_args, NextPromiseFactory next_promise_factory) {
+  // Stats object to update.
+  RefCountedPtr<GrpcLbClientStats> client_stats;
```

Comment:
Wondering why `RefCountedPtr` is preferred over the `std::` smart pointers?

Reply:
And why the original version does need a smart pointer?

Reply:
`RefCountedPtr` maintains a refcount that's always internally held - we know that if we have an object pointer we can manipulate the reference count. This is not true of `std::shared_ptr` - it supports having the refcount in a different region of memory, and so other tricks need to be made to colocate the refcount (if that's desirable, and for us it often is). Additionally, `std::shared_ptr` always supports `std::weak_ptr` which in turn adds some additional overhead to every `std::shared_ptr` - `RefCountedPtr` has no such affordances.